uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantMockMvcTests.java

```diff
@@ -84,6 +89,85 @@ void default_zone_jwt_grant() throws Exception {
             .andExpect(jsonPath("$.access_token").isNotEmpty());
     }
 
+    @Test
+    void non_default_zone_jwt_grant_user_update() throws Exception {
+        BaseClientDetails targetZoneClient = new BaseClientDetails(generator.generate(), "", "openid", "password", null);
+        targetZoneClient.setClientSecret(SECRET);
+        String subdomain = generator.generate().toLowerCase();
+        IdentityZone targetZone = MockMvcUtils.createOtherIdentityZoneAndReturnResult(subdomain,
+                mockMvc,
+                webApplicationContext,
+                targetZoneClient,
+                false, IdentityZoneHolder.getCurrentZoneId()).getIdentityZone();
+        ScimUser targetZoneUser = createUser(targetZone);
+
+        String originZoneOriginKey = createProvider(targetZone, getTokenVerificationKey(originZone.getIdentityZone()));
+
+        //Check for internal User
+        String targetZoneIdToken = getUaaIdToken(targetZone, targetZoneClient, targetZoneUser);
+        String accessTokenForTargetZoneUser = performJWTBearerGrantForJWT(targetZone, targetZoneIdToken);
+
+        //Verify JWT Bearer did not change values of internal User
+        ScimUser targetUserAfterGrant = getScimUser(targetZoneUser.getUserName(), OriginKeys.UAA, targetZone.getId());
+        assertEquals(targetZoneUser.getUserName(), targetUserAfterGrant.getUserName());
+        assertEquals(targetZoneUser.getExternalId(), targetUserAfterGrant.getExternalId());
+
+        //Check for user of registered IdP
+        String originZoneIdToken = getUaaIdToken(originZone.getIdentityZone(), originClient, originUser);
+        String accessTokenForOriginZoneUser = performJWTBearerGrantForJWT(targetZone, originZoneIdToken);
+        Map<String, Object> originUserClaims = JwtTokenUtils.getClaimsForToken(accessTokenForOriginZoneUser);
+
+        //Verify values for new shadow user set
+        ScimUser originShadowUser = getScimUser(originUser.getEmails().get(0).getValue(), originZoneOriginKey, targetZone.getId());
```

Comment:
Could we rename this to just "shadowUser" perhaps? Since there is only one shadow user involved. Previously defined vars like "originUser" means "a user in the origin zone"; "originClient" means "client in origin zone". Based on that semantic, now "originShadowUser" might mean "shadow user in origin zone", which is not true, right? It actually means the "originUser's" shadow user in the target zone.

Reply:
Yes, the latter one is correct. It is the shadow user that exists in the target zone and represents the internal user from the origin zone (which is the idp in this case). Sure we can rename that variable.